### DISCUSSION TITLE
Signing artifacts and gitlab package registry

### DISCUSSION THREAD

**Discussion by grzesuav** [goreleaser/goreleaser#4702] [Q&A] [answered]
Hi,

I had a config which is publishing to gitlab package registry - 
```yaml
gitlab_urls:
  use_job_token: true
  use_package_registry: true
```
which is working link https://gitlab.com/metacontroller/test-sem-rel/-/packages/23684709 - however it still publish on release page, is it expected ? While reading https://goreleaser.com/scm/gitlab/#generic-package-registry I was under impression it would do that instead of using gitlab release page ?

Second question, I enabled `docker_sign` with `ko` -
```yaml
docker_signs:
  - cmd: cosign
    args:
      - "sign"
      - "--upload=false"
      - "${artifact}"
      - "--yes" # needed on cosign 2.0.0+
    artifacts: all
```
using gitlab job `id_tokens` - (https://docs.gitlab.com/ee/ci/yaml/signing_examples.html#sign-or-verify-container-images-and-build-artifacts-by-using-cosign) - which works pretty neat, however while I was following docs - https://goreleaser.com/customization/docker_sign/ - (copy from docs)
```yaml
# Command line arguments for the command
    #
    # Default: ["sign", "--key=cosign.key", "${artifact}@${digest}", "--yes"]
    # Templates: allowed
    args:
    - "sign"
    - "--key=cosign.key"
    - "--upload=false"
    - "${artifact}"
    - "--yes" # needed on cosign 2.0.0+
```
it implies that `"${artifact}@${digest}"` is `image_name:image_digest`, however when I tried that, it errored : 
```shell
release failed after 1m23s               error=signing docker images: failed to publish artifacts: sign: cosign failed: exit status 1: Generating ephemeral keys...
Retrieving signed certificate...
Successfully verified SCT...
Error: signing [registry.gitlab.com/metacontroller/test-sem-rel/commanda@sha256:6ce2e017d73bf84905fc341660fdafe5531725d97012174884780d5eb63009f7@sha256:6ce2e017d73bf84905fc341660fdafe5531725d97012174884780d5eb63009f7]: parsing reference: could not parse reference: registry.gitlab.com/metacontroller/test-sem-rel/commanda@sha256:6ce2e017d73bf84905fc341660fdafe5531725d97012174884780d5eb63009f7@sha256:6ce2e017d73bf84905fc341660fdafe5531725d97012174884780d5eb63009f7
``` 

when I switched to just use `${artifact}`- https://gitlab.com/metacontroller/test-sem-rel/-/blob/main/.goreleaser.yaml?ref_type=heads#L196

it started to work and resolve image+digest - 
```shell
 signing docker images
      • signing                                      cmd=cosign artifact=registry.gitlab.com/metacontroller/test-sem-rel/commanda@sha256:bb41d1f8de551eeac77137a2d5e49667a01e83f8e5fff14a27e4fc477fcc400c
      • signing                                      cmd=cosign artifact=registry.gitlab.com/metacontroller/test-sem-rel/commandb@sha256:72fc7c39bbb11b60a5ca2236d85224a954083ef0bca170d213c244fa9c7c70d1
      • signing                                      cmd=cosign artifact=registry.gitlab.com/metacontroller/test-sem-rel/commanda@sha256:0914fb54cd8beb2b1d998a9df39efbe0ec25097be5fbc721e0565667dd74354c
      • signing                                      cmd=cosign artifact=registry.gitlab.com/metacontroller/test-sem-rel/commandb@sha256:28e56bba2081ba2032bdf2798719bc646f74421c5108a8d6bf1295ea38b77b16
      • took: 5s
    • scm releases
```
so it seems `artifact` already contain digest ?

Last thing, I created repo with example golang flow (with two binaries, each with binaries/packages/signing packages/docker images with two flavours, signed) - https://gitlab.com/metacontroller/test-sem-rel which is actually pretty  good visualization of end 2 end flow using goreleaser, if you think is it valuable I could refactor it and move/donate/make it a pr to https://gitlab.com/goreleaser/example, what do you think ?

**Comment by caarlos0** [comment, marked as answer]
ah, yeah, that is a bad example indeed, will fix it


about the repo, if you wanna transfer it to the goreleaser org, I would be more than happy to accept it.

we have a bunch of different examples in the github org, but only that single, simple one, in gitlab

**Reply by grzesuav** [reply]
ok, probably need to do some cleaning, will take a look after kubecon happening next week. Feel free to ping me in case I forgot :)

I would be happy transferring the example, thanks for the great tool

**Reply by grzesuav** [reply]
hi @caarlos0 , can you take a look on https://gitlab.com/metacontroller/test-sem-rel and see if it is ready for transfer ?

**Reply by grzesuav** [reply]
one thing to clarify is how to rename the project

**Reply by grzesuav** [reply]
@caarlos0 any suggestion ?

**Reply by grzesuav** [reply]
Also, to transfer a project : 

```
A project’s repository name defines its URL (the one you use to access the project via a browser) and its place on the file disk where GitLab is installed. [Learn more.](https://gitlab.com/help/user/project/working_with_projects#rename-a-repository)

When you transfer your project to a group, you can easily manage multiple projects, view usage quotas for storage, pipeline minutes, and users, and start a trial or upgrade to a paid tier.

Don't have a group? [Create one](https://gitlab.com/groups/new)

Things to be aware of before transferring:

Be careful. Changing the project's namespace can have unintended side effects.
You can only transfer the project to namespaces you manage.
You will need to update your local repositories to point to the new location.
Project visibility level will be changed to match namespace rules when transferring to a group.
```
I can add you as a maintainer and you could transfer it, how does it sounds ?

**Reply by caarlos0** [reply]
hey, sorry for the delay

regarding naming, we usually name the examples `goreleaser-example-<something>`
regarding adding me as a maintainer, that sounds good as well

**Reply by grzesuav** [reply]
updated the name, did a test release, send you an invite to https://gitlab.com/metacontroller/goreleaser-example-multi-binary-multi-image

**Reply by grzesuav** [reply]
@caarlos0 are you able to transfer the project ? Or I need to grant owner access?

**Reply by caarlos0** [reply]
hey sorry for the delay... i couldn't find the option to transfer, not sure if i'm blind or its not there... 

anyway, to prevent back and forth, i cloned the repo, created another repo with the same name under goreleaser's org, changed remotes, and pushed 

https://gitlab.com/goreleaser/goreleaser-example-multi-binary-multi-image

I think this should be enough?
Will also give you access to it!

thanks!

**Reply by grzesuav** [reply]
ah, that great, did not think of that ;)

will try to make some silly change and do a test release to double check it is working after moving

**Reply by grzesuav** [reply]
ah, I see you did a release and it is publishing everything, packages, container images and helm. So everything is ok

**Reply by caarlos0** [reply]
noice! thanks again for all the help!
